File: leetcode/150-evaluate-reverse-polish-notation.js
// https://leetcode.com/problems/evaluate-reverse-polish-notation/ (top-interview-questions)

/**
 * @param {string[]} tokens
 * @return {number}
 */
var evalRPN = function(tokens) {
  const stack = [];

  const operations = {
    '+': (a, b) => (a + b),
    '-': (a, b) => (a - b),
    '*': (a, b) => (a * b),
    '/': (a, b) => (Math.trunc(a / b))
  }

  let a;
  let b;

  for (const token of tokens) {
    if (operations.hasOwnProperty(token)) {
      b = stack.pop();
      a = stack.pop();
      stack.push(operations[token](a, b));
    } else {
      stack.push(Number(token));
    }
  }

  return stack.pop();
};

File: leetcode/150-evaluate-reverse-polish-notation.py
# https://leetcode.com/problems/evaluate-reverse-polish-notation/ (top-interview-questions)

from typing import List

class Solution:
    def evalRPN(self, tokens: List[str]) -> int:
        resultStack = list()
        for token in tokens:
            if token == '+':
                resultStack.append(resultStack.pop() + resultStack.pop())
            elif token == '-':
                num2 = resultStack.pop()
                num1 = resultStack.pop()
                resultStack.append(num1 - num2)
            elif token == '*':
                resultStack.append(resultStack.pop() * resultStack.pop())
            elif token == '/':
                num2 = resultStack.pop()
                num1 = resultStack.pop()
                resultStack.append(int(num1 / num2))
            else:
                resultStack.append(int(token))
        return resultStack.pop()
